sample-plugin-candlestick-chart-new/src/CandlestickChart.js

```diff
@@ -0,0 +1,153 @@
+import React from "react";
+import PropTypes from "prop-types";
+
+import { format } from "d3-format";
+import { ChartCanvas, Chart } from "react-stockcharts";
+import { CandlestickSeries } from "react-stockcharts/lib/series";
+import { XAxis, YAxis } from "react-stockcharts/lib/axes";
+import {
+	CrossHairCursor,
+	EdgeIndicator,
+	MouseCoordinateY,
+} from "react-stockcharts/lib/coordinates";
+
+import { discontinuousTimeScaleProvider } from "react-stockcharts/lib/scale";
+import { OHLCTooltip } from "react-stockcharts/lib/tooltip";
+import { ema, stochasticOscillator, bollingerBand } from "react-stockcharts/lib/indicator";
+import { fitWidth } from "react-stockcharts/lib/helper";
+import { last } from "react-stockcharts/lib/utils";
+
+class CandlestickChart extends React.Component {
+	render() {
+		const { type, data: initialData, width, ratio } = this.props;
+		const height = 500;
+		const margin = { left: 70, right: 70, top: 20, bottom: 30 };
+		const gridHeight = height - margin.top - margin.bottom;
+		const gridWidth = width - margin.left - margin.right;
+		const showGrid = true;
+		const yGrid = showGrid ? { innerTickSize: -1 * gridWidth, tickStrokeOpacity: 0.2 } : {};
+		const xGrid = showGrid ? { innerTickSize: -1 * gridHeight, tickStrokeOpacity: 0.2 } : {};
+
+		// START helper functions for additional feature support
+		const ema20 = ema()
+			.id(0)
+			.options({ windowSize: 20 })
+			.merge((d, c) => {d.ema20 = c;})
+			.accessor(d => d.ema20);
+
+		const ema50 = ema()
+			.id(2)
+			.options({ windowSize: 50 })
+			.merge((d, c) => {d.ema50 = c;})
+			.accessor(d => d.ema50);
+
+		const slowSTO = stochasticOscillator()
+			.options({ windowSize: 14, kWindowSize: 3 })
+			.merge((d, c) => {d.slowSTO = c;})
+			.accessor(d => d.slowSTO);
+
+		const fastSTO = stochasticOscillator()
+			.options({ windowSize: 14, kWindowSize: 1 })
+			.merge((d, c) => {d.fastSTO = c;})
+			.accessor(d => d.fastSTO);
+
+		const fullSTO = stochasticOscillator()
+			.options({ windowSize: 14, kWindowSize: 3, dWindowSize: 4 })
+			.merge((d, c) => {d.fullSTO = c;})
+			.accessor(d => d.fullSTO);
+
+		const bb = bollingerBand()
+			.merge((d, c) => {d.bb = c;})
+			.accessor(d => d.bb);
+		// END helper functions for additional feature support
+
+		const calculatedData = bb(ema20(ema50(slowSTO(fastSTO(fullSTO(initialData))))));
+		const xScaleProvider = discontinuousTimeScaleProvider.inputDateAccessor(d => d.date);
+
+		const {
+			data,
+			xScale,
+			xAccessor,
+			displayXAccessor,
+		} = xScaleProvider(calculatedData);
+
+		const start = xAccessor(last(data));
+		const end = xAccessor(data[Math.max(0, data.length - 150)]);
+		const xExtents = [start, end];
```

Comment:
Once you convert to a functional  component, most of this logic should probably be put into a `useMemo` hook